[WPE] Favicon database support

aperezdc opened the pull request:
#### a79ea53815c4f69982c3eab9a3d97c797dcd5b13
<pre>
[WPE] Favicon database support
<a href="https://bugs.webkit.org/show_bug.cgi?id=296927">https://bugs.webkit.org/show_bug.cgi?id=296927</a>

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp

* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/IconDatabase.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(_WebKitFavicon::width const):
(_WebKitFavicon::height const):
(_WebKitFavicon::ensureBytes const):
(_WebKitFavicon::stride const):
(_WebKitFavicon::ensureBytes):
(_WebKitFavicon::create):
(_WebKitFavicon::_WebKitFavicon):
(webkit_favicon_ref):
(webkit_favicon_unref):
(webkit_favicon_get_width):
(webkit_favicon_get_height):
(webkit_favicon_get_stride):
(webkit_favicon_get_bytes):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabasePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(gotFaviconCallback):
(webkitWebViewConstructed):
(webkitWebViewGetProperty):
(webkitWebViewDispose):
(webkit_web_view_class_init):
(webkitWebViewLoadChanged):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.h.in:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(testFaviconDatabaseGetFavicon):
(afterAll):
* Tools/TestWebKitAPI/glib/PlatformWPE.cmake:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a79ea53815c4f69982c3eab9a3d97c797dcd5b13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123800 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43515 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130579 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75950 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44239 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52109 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94163 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62482 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126753 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110753 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74754 "Found 1 new API test failure: WPE/TestWebKitFaviconDatabase:/webkit/WebKitFaviconDatabase/get-favicon (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28913 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74062 "Hash a79ea538 for PR 49452 does not build (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104969 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133275 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50752 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38651 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102624 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51127 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106972 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102454 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47800 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26056 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47602 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50606 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56370 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50080 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53426 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -37,12 +37,124 @@
+    explicit _WebKitFavicon(ImageType&& image)
+        : m_image(WTFMove(image))
+    {
+    }
```
This could be private.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
```diff
@@ -36,6 +36,32 @@
 G_BEGIN_DECLS
 
+#if PLATFORM(WPE) && ENABLE(2022_GLIB_API)
+#define WEBKIT_TYPE_FAVICON (webkit_favicon_get_type())
+typedef struct _WebKitFavicon WebKitFavicon;
```
I suppose this could be called `WPEImage` or something similar if we want to reuse it in more places.

aperezdc replied:
@carlosgcampos Any thoughts on this?

aperezdc replied:
Actually, I am going to call it `WebKitImage` and make it a `GObject` subclass, that way we will be able to make it implement the `GIcon` interface (and maybe `GLoadableIcon` as well). Plus, then we could reuse it to implement `webkit_web_view_get_snapshot()` for WPE in the future.

aperezdc commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
```diff
@@ -260,7 +268,8 @@ static void testFaviconDatabaseInitialization(FaviconDatabaseTest* test, gconstp
     test->getFaviconForPageURIAndWaitUntilReady(kServer->getURIForPath("/foo").data());
-    g_assert_null(test->m_favicon);
+    // FIXME: This check fails.
+    // g_assert_null(test->m_favicon);
```
This is failing and I have not had the chance yet to investigate or to try using the API to check that things work. That's the reason why the PR is marked as a draft.

aperezdc replied:
Fixed with an early null return when trying to get a favicon from the database fails; before the code was returning a `WPEFavicon` that wrapped a null image.

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -280,6 +394,22 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+    auto image = adoptGRef(static_cast<cairo_surface_t*>(g_task_propagate_pointer(G_TASK(result), error)));
+#endif
+
+    return WebKitFavicon::create(WTFMove(image));
```
AFAIU we will always get a RGBA8888 image here, and that's the reason why I have not added a `webkit_favicon_get_format()` getter.

If we might get something else here, we have two options: a.) always convert to a given format before creating the `WebKitFavicon`, and RGBA8888 should be fine for any kind of favicon IMO; or b.) add the getter for the pixel format and let client code deal with conversions and who knows how many possible formats. One last option could be to do both: add `webkit_favicon_get_format()` _and_ a `webkit_favicon_as_rgba8888()` that converts if needed.

Thoughts welcome!

aperezdc replied:
Correction: we currently get BGRA8888 on little-endian.

spenap commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -37,12 +37,126 @@
+    inline unsigned height() const { return cairo_image_surface_get_height(m_image.get()); }
+    inline unsigned stride() const { return cairo_image_surface_get_stride(m_image.get()); }
+
+    GBytes* ensureBytes()
```
I understand the Cairo API takes non-const pointers to `m_image` so we can't make this method `const` unless we also make the image mutable, right?

nikolaszimmermann replied:
I wouldn't add new Cairo code at all.

aperezdc replied:
> I wouldn't add new Cairo code at all.

We are still going to need the Cairo code path, because WPE can be built with `USE_SKIA=OFF`.

spenap reviewed (commented):
Changes look very good, very neat, thanks!

aperezdc committed:
[WPE] Favicon database support
https://bugs.webkit.org/show_bug.cgi?id=296927

Reviewed by NOBODY (OOPS!).

Explanation of why this fixes the bug (OOPS!).

Tests: Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp

* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/IconDatabase.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
(_WebKitFavicon::width const):
(_WebKitFavicon::height const):
(_WebKitFavicon::ensureBytes const):
(_WebKitFavicon::stride const):
(_WebKitFavicon::ensureBytes):
(_WebKitFavicon::create):
(_WebKitFavicon::_WebKitFavicon):
(webkit_favicon_ref):
(webkit_favicon_unref):
(webkit_favicon_get_width):
(webkit_favicon_get_height):
(webkit_favicon_get_stride):
(webkit_favicon_get_bytes):
(webkit_favicon_database_get_favicon_finish):
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabasePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitIconLoadingClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(gotFaviconCallback):
(webkitWebViewConstructed):
(webkitWebViewGetProperty):
(webkitWebViewDispose):
(webkit_web_view_class_init):
(webkitWebViewLoadChanged):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebsiteDataManager.h.in:
* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitFaviconDatabase.cpp:
(testFaviconDatabaseGetFavicon):
(afterAll):
* Tools/TestWebKitAPI/glib/PlatformWPE.cmake:

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -37,12 +37,124 @@
+
+WebKitFavicon* webkit_favicon_ref(WebKitFavicon* favicon)
+{
+    g_atomic_int_inc(&favicon->referenceCount);
```
I don't think there is a practical difference so feel free to ignore, but GLib specifically has reference counting APIs now:

`g_atomic_ref_count_inc()`

aperezdc replied:
Good point, the API you suggest is available since GLib 2.58 and currently we require 2.70, so we can use it. Thanks!

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.cpp:
```diff
@@ -280,6 +394,27 @@ cairo_surface_t* webkit_favicon_database_get_favicon_finish(WebKitFaviconDatabas
+        return WebKitFavicon::create(WTFMove(image));
+
+    if (error && !*error)
+        g_set_error_literal(error, WEBKIT_FAVICON_DATABASE_ERROR, WEBKIT_FAVICON_DATABASE_ERROR_FAVICON_UNKNOWN, _("Failed to create image"));
```
`g_set_error_literal` itself handles if `error` is `NULL`, and its intentionally a critical to pass an already set error.

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitFaviconDatabase.h.in:
```diff
@@ -96,6 +124,13 @@ webkit_favicon_database_get_favicon_finish (WebKitFaviconDatabase *database,
 
+#if PLATFORM(WPE) && ENABLE(2022_GLIB_API)
+WEBKIT_API WebKitFavicon *
+webkit_favicon_database_get_favicon_finish (WebKitFaviconDatabase *database,
+                                            GAsyncResult          *result,
+                                            GError               **error);
```
```suggestion
webkit_favicon_database_get_favicon_finish (WebKitFaviconDatabase  *database,
                                            GAsyncResult           *result,
                                            GError                **error);
```
